File: src/utils/encryption.js
import CryptoJS from 'crypto-js';

/**
 * Frontend decryption utilities for handling encrypted questions
 */
class QuestionDecryption {
  
  /**
   * Decrypt a single encrypted question
   * @param {Object} encryptedQuestion - The encrypted question object
   * @param {string} keyHex - The encryption key as hex string
   * @returns {Object} - Decrypted question object
   */
  static decryptQuestion(encryptedQuestion, keyHex) {
    try {
      const encryptedData = encryptedQuestion.encrypted_data;
      
      // Convert hex key to WordArray
      const key = CryptoJS.enc.Hex.parse(keyHex);
      
      // Decode base64 encrypted data
      const encryptedBytes = CryptoJS.enc.Base64.parse(encryptedData);
      
      // Extract IV (first 16 bytes) and encrypted content
      const iv = CryptoJS.lib.WordArray.create(encryptedBytes.words.slice(0, 4));
      const encrypted = CryptoJS.lib.WordArray.create(encryptedBytes.words.slice(4));
      
      // Decrypt using AES-CBC
      const decrypted = CryptoJS.AES.decrypt(
        { ciphertext: encrypted },
        key,
        { 
          iv: iv, 
          mode: CryptoJS.mode.CBC, 
          padding: CryptoJS.pad.Pkcs7 
        }
      );
      
      // Convert to string and parse JSON
      const decryptedText = decrypted.toString(CryptoJS.enc.Utf8);
      
      if (!decryptedText) {
        throw new Error('Decryption resulted in empty string - invalid key or corrupted data');
      }
      
      const questionData = JSON.parse(decryptedText);
      
      // Return complete question object
      return {
        questionID: encryptedQuestion.questionID,
        topic: encryptedQuestion.topic,
        question: questionData.question,
        options: questionData.options,
        correctAnswer: questionData.correctAnswer,
        datetime: encryptedQuestion.datetime
      };
      
    } catch (error) {
      console.error('Decryption error:', error);
      throw new Error(`Failed to decrypt question: ${error.message}`);
    }
  }
  
  /**
   * Decrypt the current question based on index
   * @param {Array} encryptedQuestions - Array of encrypted questions
   * @param {number} currentIndex - Current question index
   * @param {string} keyHex - The encryption key as hex string
   * @returns {Object|null} - Decrypted question or null if invalid index
   */
  static decryptCurrentQuestion(encryptedQuestions, currentIndex, keyHex) {
    if (currentIndex >= 0 && currentIndex < encryptedQuestions.length) {
      return this.decryptQuestion(encryptedQuestions[currentIndex], keyHex);
    }
    return null;
  }
  
  /**
   * Validate encryption key by attempting to decrypt first question
   * @param {Array} encryptedQuestions - Array of encrypted questions
   * @param {string} keyHex - The encryption key as hex string
   * @returns {boolean} - True if key is valid
   */
  static validateKey(encryptedQuestions, keyHex) {
    if (!encryptedQuestions || encryptedQuestions.length === 0) {
      return false;
    }
    
    try {
      this.decryptQuestion(encryptedQuestions[0], keyHex);
      return true;
    } catch (error) {
      console.warn('Key validation failed:', error.message);
      return false;
    }
  }
  
  /**
   * Decrypt multiple questions (use sparingly - decrypt on demand is preferred)
   * @param {Array} encryptedQuestions - Array of encrypted questions
   * @param {string} keyHex - The encryption key as hex string
   * @returns {Array} - Array of decrypted questions
   */
  static decryptAllQuestions(encryptedQuestions, keyHex) {
    const decryptedQuestions = [];
    
    for (let i = 0; i < encryptedQuestions.length; i++) {
      try {
        const decrypted = this.decryptQuestion(encryptedQuestions[i], keyHex);
        decryptedQuestions.push(decrypted);
      } catch (error) {
        console.error(`Failed to decrypt question ${i}:`, error);
        // Add placeholder for failed decryption to maintain array indices
        decryptedQuestions.push({
          questionID: encryptedQuestions[i].questionID,
          topic: encryptedQuestions[i].topic,
          question: 'Error: Failed to decrypt question',
          options: ['Error loading options'],
          correctAnswer: 'Error',
          error: true
        });
      }
    }
    
    return decryptedQuestions;
  }
  
  /**
   * Get question metadata without decrypting sensitive content
   * @param {Object} encryptedQuestion - The encrypted question object
   * @returns {Object} - Question metadata
   */
  static getQuestionMetadata(encryptedQuestion) {
    return {
      questionID: encryptedQuestion.questionID,
      topic: encryptedQuestion.topic,
      datetime: encryptedQuestion.datetime,
      hasEncryptedData: !!encryptedQuestion.encrypted_data
    };
  }
}

/**
 * Answer encryption utilities for secure answer submission
 */
class AnswerEncryption {
  
  /**
   * Encrypt an answer before submission
   * @param {Object} answerData - The answer data to encrypt
   * @param {string} keyHex - The encryption key as hex string
   * @returns {string} - Encrypted answer as base64 string
   */
  static encryptAnswer(answerData, keyHex) {
    try {
      // Convert answer data to JSON
      const jsonData = JSON.stringify(answerData);
      
      // Convert hex key to WordArray
      const key = CryptoJS.enc.Hex.parse(keyHex);
      
      // Generate random IV
      const iv = CryptoJS.lib.WordArray.random(16);
      
      // Encrypt using AES-CBC
      const encrypted = CryptoJS.AES.encrypt(jsonData, key, {
        iv: iv,
        mode: CryptoJS.mode.CBC,
        padding: CryptoJS.pad.Pkcs7
      });
      
      // Combine IV and encrypted data
      const combined = iv.concat(encrypted.ciphertext);
      
      // Return as base64 string
      return CryptoJS.enc.Base64.stringify(combined);
      
    } catch (error) {
      console.error('Answer encryption error:', error);
      throw new Error(`Failed to encrypt answer: ${error.message}`);
    }
  }
  
  /**
   * Encrypt multiple answers for batch submission
   * @param {Array} answers - Array of answer objects
   * @param {string} keyHex - The encryption key as hex string
   * @returns {Array} - Array of encrypted answers
   */
  static encryptAnswers(answers, keyHex) {
    return answers.map((answer, index) => {
      try {
        return {
          index: index,
          encrypted_answer: this.encryptAnswer(answer, keyHex),
          questionID: answer.questionID,
          timestamp: answer.timestamp || new Date().toISOString()
        };
      } catch (error) {
        console.error(`Failed to encrypt answer ${index}:`, error);
        return {
          index: index,
          error: error.message,
          questionID: answer.questionID,
          timestamp: answer.timestamp || new Date().toISOString()
        };
      }
    });
  }
}

export { QuestionDecryption, AnswerEncryption };

File: backend/getAllQuestionsForTest/encryption.py
import base64
import json
import hashlib
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad, unpad
import secrets

class QuestionEncryption:
    """
    Handles encryption and decryption of questions for secure bulk delivery
    """
    
    @staticmethod
    def generate_test_key(test_id, candidate_name):
        """
        Generate a unique encryption key for each test session
        """
        # Create a deterministic but secure key based on test and candidate
        key_material = f"{test_id}_{candidate_name}_{secrets.token_hex(16)}"
        key = hashlib.sha256(key_material.encode()).digest()
        return key
    
    @staticmethod
    def encrypt_question(question_data, key):
        """
        Encrypt a single question's sensitive data
        """
        try:
            # Extract sensitive data to encrypt
            sensitive_data = {
                'question': question_data.get('question', ''),
                'options': question_data.get('options', []),
                'correctAnswer': question_data.get('correctAnswer', '')
            }
            
            # Convert to JSON string
            json_data = json.dumps(sensitive_data, ensure_ascii=False)
            
            # Generate random IV for each encryption
            iv = get_random_bytes(16)
            
            # Create cipher and encrypt
            cipher = AES.new(key, AES.MODE_CBC, iv)
            padded_data = pad(json_data.encode('utf-8'), AES.block_size)
            encrypted_data = cipher.encrypt(padded_data)
            
            # Combine IV and encrypted data, then base64 encode
            encrypted_payload = base64.b64encode(iv + encrypted_data).decode('utf-8')
            
            # Return question with encrypted sensitive data
            encrypted_question = {
                'questionID': question_data.get('questionID'),
                'topic': question_data.get('topic', '__NO_TOPIC__'),
                'encrypted_data': encrypted_payload,
                'datetime': question_data.get('datetime')
            }
            
            return encrypted_question
            
        except Exception as e:
            print(f"Encryption error: {str(e)}")
            raise Exception(f"Failed to encrypt question: {str(e)}")
    
    @staticmethod
    def encrypt_questions_bulk(questions, key):
        """
        Encrypt multiple questions efficiently
        """
        encrypted_questions = []
        
        for question in questions:
            try:
                encrypted_question = QuestionEncryption.encrypt_question(question, key)
                encrypted_questions.append(encrypted_question)
            except Exception as e:
                print(f"Failed to encrypt question {question.get('questionID', 'unknown')}: {str(e)}")
                # Skip failed encryptions rather than failing entire batch
                continue
        
        return encrypted_questions
    
    @staticmethod
    def create_key_hint(key):
        """
        Create a hint for key verification (first 8 chars of key hash)
        """
        key_hash = hashlib.sha256(key).hexdigest()
        return key_hash[:8]

# JavaScript equivalent for frontend (to be used in React)
FRONTEND_DECRYPTION_CODE = '''
// Frontend decryption utilities (add to utils/encryption.js)
import CryptoJS from 'crypto-js';

class QuestionDecryption {
  static generateTestKey(testId, candidateName, keyHint) {
    // This would need to match the backend key generation
    // For security, the key should be provided by backend or derived securely
    return keyHint; // Placeholder - implement secure key derivation
  }
  
  static decryptQuestion(encryptedQuestion, key) {
    try {
      const encryptedData = encryptedQuestion.encrypted_data;
      
      // Decode base64
      const encryptedBytes = CryptoJS.enc.Base64.parse(encryptedData);
      
      // Extract IV (first 16 bytes) and encrypted data
      const iv = CryptoJS.lib.WordArray.create(encryptedBytes.words.slice(0, 4));
      const encrypted = CryptoJS.lib.WordArray.create(encryptedBytes.words.slice(4));
      
      // Decrypt
      const decrypted = CryptoJS.AES.decrypt(
        { ciphertext: encrypted },
        CryptoJS.enc.Hex.parse(key),
        { iv: iv, mode: CryptoJS.mode.CBC, padding: CryptoJS.pad.Pkcs7 }
      );
      
      // Convert to string and parse JSON
      const decryptedText = decrypted.toString(CryptoJS.enc.Utf8);
      const questionData = JSON.parse(decryptedText);
      
      // Return complete question object
      return {
        questionID: encryptedQuestion.questionID,
        topic: encryptedQuestion.topic,
        question: questionData.question,
        options: questionData.options,
        correctAnswer: questionData.correctAnswer,
        datetime: encryptedQuestion.datetime
      };
      
    } catch (error) {
      console.error('Decryption error:', error);
      throw new Error('Failed to decrypt question');
    }
  }
  
  static decryptCurrentQuestion(encryptedQuestions, currentIndex, key) {
    if (currentIndex >= 0 && currentIndex < encryptedQuestions.length) {
      return this.decryptQuestion(encryptedQuestions[currentIndex], key);
    }
    return null;
  }
}

export default QuestionDecryption;
'''